--- a/common/autoboot.c	2018-12-18 19:36:45.905712000 +0900
+++ b/common/autoboot.c	2019-11-21 13:39:46.895141150 +0900
@@ -12,9 +12,12 @@
 #include <fdtdec.h>
 #include <menu.h>
 #include <post.h>
+#include <asm/gpio.h>
 
 DECLARE_GLOBAL_DATA_PTR;
 
+#define CONFIG_ZERO_BOOTDELAY_CHECK
+
 #define MAX_DELAY_STOP_STR 32
 
 #ifndef DEBUG_BOOTKEYS
@@ -270,7 +273,19 @@ const char *bootdelay_process(void)
 		s = getenv("altbootcmd");
 	} else
 #endif /* CONFIG_BOOTCOUNT_LIMIT */
+	{
+		if (gpio_get_value(GPIO_GP_5_5) || strcmp(getenv("boot_type"), "normal")) {
+			setenv("part", "1:3");
+			setenv("bootcmd", "run bootcmd_emmc");
+		} else {
+			if (strcmp(getenv("kernel_bank"), "A") == 0) {
+				setenv("part", "1:1");
+			} else {
+				setenv("part", "1:2");
+			}
+		}
 		s = getenv("bootcmd");
+	}
 
 	process_fdt_options(gd->fdt_blob);
 	stored_bootdelay = bootdelay;
