--- a/board/renesas/salvator-x/salvator-x.c	2018-09-13 17:01:21.577064000 +0900
+++ b/board/renesas/salvator-x/salvator-x.c	2019-06-27 11:40:38.107635356 +0900
@@ -39,7 +39,11 @@ DECLARE_GLOBAL_DATA_PTR;
 #define GPIO3_MSTP909	(1 << 9)	/* GPIO3: includes assignment of SDHI3 */
 #define GPIO5_MSTP907	(1 << 7)	/* GPIO5: includes assignment of SDHI0/1/2 */
 #define GPIO6_MSTP906	(1 << 6)	/* GPIO6: includes assignment of USB */
-
+#define GPIO1_MSTP911	(1 << 11)	/* GPIO1 */
+#define GPIO4_MSTP908	(1 << 8)	/* GPIO4 */
+#define GPIO7_MSTP905	(1 << 5)	/* GPIO7 */
+#define PWM_MSTP523	(1 << 23)	/* PWM */
+void pwm_bl_ctrl(void);
 #define SD0CKCR		0xE6150074
 #define SD1CKCR		0xE6150078
 #define SD2CKCR		0xE6150268
@@ -57,8 +61,10 @@ int board_early_init_f(void)
 	mstp_clrbits_le32(SMSTPCR8, SMSTPCR8, ETHERAVB_MSTP812);
 	/* eMMC */
 	mstp_clrbits_le32(SMSTPCR3, SMSTPCR3, SD1_MSTP313 | SD2_MSTP312);
-	/* SDHI0, 3 */
-	mstp_clrbits_le32(SMSTPCR3, SMSTPCR3, SD0_MSTP314 | SD3_MSTP311);
+	/* SDHI0 */
+	mstp_clrbits_le32(SMSTPCR3, SMSTPCR3, SD0_MSTP314);
+	/* PWM */
+	mstp_clrbits_le32(SMSTPCR5, SMSTPCR5, PWM_MSTP523);
 #if defined(CONFIG_R8A7795)
 	if (rcar_is_legacy()) {
 		mstp_clrbits_le32(SMSTPCR9, SMSTPCR9,
@@ -71,8 +77,8 @@ int board_early_init_f(void)
 	}
 #elif defined(CONFIG_R8A7796X)
 	mstp_clrbits_le32(SMSTPCR9, SMSTPCR9,
-			  GPIO2_MSTP910 | GPIO3_MSTP909 |
-			  GPIO5_MSTP907 | GPIO6_MSTP906);
+			  GPIO2_MSTP910 | GPIO3_MSTP909 | GPIO1_MSTP911 | GPIO4_MSTP908 |
+			  GPIO5_MSTP907 | GPIO6_MSTP906 | GPIO7_MSTP905);
 #endif
 	freq = rcar_get_sdhi_config_clk();
 	writel(freq, SD0CKCR);
@@ -94,8 +100,8 @@ void board_cleanup_preboot_os(void)
 	mstp_setbits_le32(SMSTPCR8, SMSTPCR8, ETHERAVB_MSTP812);
 	/* eMMC */
 	mstp_setbits_le32(SMSTPCR3, SMSTPCR3, SD1_MSTP313 | SD2_MSTP312);
-	/* SDHI0, 3 */
-	mstp_setbits_le32(SMSTPCR3, SMSTPCR3, SD0_MSTP314 | SD3_MSTP311);
+	/* SDHI0 */
+	mstp_setbits_le32(SMSTPCR3, SMSTPCR3, SD0_MSTP314);
 #if defined(CONFIG_SYS_I2C) && defined(CONFIG_SYS_I2C_SH)
 	/* DVFS for reset */
 	mstp_setbits_le32(SMSTPCR9, SMSTPCR9, DVFS_MSTP926);
@@ -234,6 +240,29 @@ int board_init(void)
 	udelay(1);
 #endif
 
+	/* MUTE ON */
+	gpio_request(GPIO_GP_6_21, NULL);
+	gpio_direction_output(GPIO_GP_6_21, 0);
+	/* DAC CS */
+	gpio_request(GPIO_GP_5_19, NULL);
+	gpio_direction_output(GPIO_GP_5_19, 1);
+	/* PANEL UPDATE REQUEST */
+	gpio_request(GPIO_GP_5_5, NULL);
+	gpio_direction_input(GPIO_GP_5_5);
+	/* MAIN LCD BACKLIGHT */
+	pwm_bl_ctrl();
+	gpio_request(GPIO_GP_2_2, NULL);
+	gpio_direction_output(GPIO_GP_2_2, 0); // BL1_ENSEC
+	gpio_request(GPIO_GP_2_3, NULL);
+	gpio_direction_output(GPIO_GP_2_3, 0); // BL0_DSCH
+	gpio_request(GPIO_GP_1_21, NULL);
+	gpio_direction_output(GPIO_GP_1_21, 0); // BL0_FREN
+	gpio_request(GPIO_GP_1_26, NULL);
+	gpio_direction_output(GPIO_GP_1_26, 0); // BL1_ENPRE
+	/* FAN CONTROL */
+	gpio_request(GPIO_GP_2_8, NULL);
+	gpio_direction_output(GPIO_GP_2_8, 1);
+
 #endif
 	return 0;
 }
@@ -417,10 +446,11 @@ int board_mmc_init(bd_t *bis)
 	gpio_request(GPIO_GFN_SD0_CD, NULL);
 	gpio_request(GPIO_GFN_SD0_WP, NULL);
 
-	gpio_request(GPIO_GP_5_2, NULL);
-	gpio_request(GPIO_GP_5_1, NULL);
-	gpio_direction_output(GPIO_GP_5_2, 1);	/* power on */
-	gpio_direction_output(GPIO_GP_5_1, 1);	/* 1: 3.3V, 0: 1.8V */
+	gpio_request(GPIO_GP_4_7, NULL);
+	gpio_direction_output(GPIO_GP_4_7, 1); /* power on */
+
+	gpio_request(GPIO_GP_5_14, NULL);
+	gpio_direction_output(GPIO_GP_5_14, 1);	/* 1: 3.3V, 0: 1.8V */
 
 	ret = sh_sdhi_init(CONFIG_SYS_SH_SDHI0_BASE, 0,
 			   SH_SDHI_QUIRK_64BIT_BUF);
@@ -439,33 +469,8 @@ int board_mmc_init(bd_t *bis)
 	gpio_request(GPIO_GFN_SD2_CLK, NULL);
 	gpio_request(GPIO_FN_SD2_CMD, NULL);
 
-	gpio_request(GPIO_GP_5_3, NULL);
-	gpio_request(GPIO_GP_5_9, NULL);
-	gpio_direction_output(GPIO_GP_5_3, 0);	/* 1: 3.3V, 0: 1.8V */
-	gpio_direction_output(GPIO_GP_5_9, 0);	/* 1: 3.3V, 0: 1.8V */
-
 	ret = sh_sdhi_init(CONFIG_SYS_SH_SDHI2_BASE, 1,
 			   SH_SDHI_QUIRK_64BIT_BUF);
-	if (ret)
-		return ret;
-
-	/* SDHI3 */
-	gpio_request(GPIO_FN_SD3_DAT0, NULL);	/* GP_4_9 */
-	gpio_request(GPIO_FN_SD3_DAT1, NULL);	/* GP_4_10 */
-	gpio_request(GPIO_FN_SD3_DAT2, NULL);	/* GP_4_11 */
-	gpio_request(GPIO_FN_SD3_DAT3, NULL);	/* GP_4_12 */
-	gpio_request(GPIO_FN_SD3_CLK, NULL);	/* GP_4_7 */
-	gpio_request(GPIO_FN_SD3_CMD, NULL);	/* GP_4_8 */
-	gpio_request(GPIO_FN_SD3_CD, NULL);	/* GP_4_15 */
-	gpio_request(GPIO_FN_SD3_WP, NULL);	/* GP_4_16 */
-
-	gpio_request(GPIO_GP_3_15, NULL);
-	gpio_request(GPIO_GP_3_14, NULL);
-	gpio_direction_output(GPIO_GP_3_15, 1);	/* power on */
-	gpio_direction_output(GPIO_GP_3_14, 1);	/* 1: 3.3V, 0: 1.8V */
-
-	ret = sh_sdhi_init(CONFIG_SYS_SH_SDHI3_BASE, 2,
-			   SH_SDHI_QUIRK_64BIT_BUF);
 #endif
 
 #endif
