--- a/drivers/usb/gadget/function/f_hid.c	2018-11-27 18:10:37.000000000 +0900
+++ b/drivers/usb/gadget/function/f_hid.c	2019-09-25 15:36:10.133271506 +0900
@@ -103,10 +103,10 @@ static struct hid_descriptor hidg_desc =
 static struct usb_endpoint_descriptor hidg_hs_in_ep_desc = {
 	.bLength		= USB_DT_ENDPOINT_SIZE,
 	.bDescriptorType	= USB_DT_ENDPOINT,
-	.bEndpointAddress	= USB_DIR_IN,
+	.bEndpointAddress	= USB_DIR_IN | 0x07,
 	.bmAttributes		= USB_ENDPOINT_XFER_INT,
 	/*.wMaxPacketSize	= DYNAMIC */
-	.bInterval		= 4, /* FIXME: Add this field in the
+	.bInterval		= 1, /* FIXME: Add this field in the
 				      * HID gadget configuration?
 				      * (struct hidg_func_descriptor)
 				      */
@@ -115,10 +115,10 @@ static struct usb_endpoint_descriptor hi
 static struct usb_endpoint_descriptor hidg_hs_out_ep_desc = {
 	.bLength		= USB_DT_ENDPOINT_SIZE,
 	.bDescriptorType	= USB_DT_ENDPOINT,
-	.bEndpointAddress	= USB_DIR_OUT,
+	.bEndpointAddress	= USB_DIR_OUT | 0x06,
 	.bmAttributes		= USB_ENDPOINT_XFER_INT,
 	/*.wMaxPacketSize	= DYNAMIC */
-	.bInterval		= 4, /* FIXME: Add this field in the
+	.bInterval		= 1, /* FIXME: Add this field in the
 				      * HID gadget configuration?
 				      * (struct hidg_func_descriptor)
 				      */
@@ -140,7 +140,7 @@ static struct usb_endpoint_descriptor hi
 	.bEndpointAddress	= USB_DIR_IN,
 	.bmAttributes		= USB_ENDPOINT_XFER_INT,
 	/*.wMaxPacketSize	= DYNAMIC */
-	.bInterval		= 10, /* FIXME: Add this field in the
+	.bInterval		= 1, /* FIXME: Add this field in the
 				       * HID gadget configuration?
 				       * (struct hidg_func_descriptor)
 				       */
@@ -152,7 +152,7 @@ static struct usb_endpoint_descriptor hi
 	.bEndpointAddress	= USB_DIR_OUT,
 	.bmAttributes		= USB_ENDPOINT_XFER_INT,
 	/*.wMaxPacketSize	= DYNAMIC */
-	.bInterval		= 10, /* FIXME: Add this field in the
+	.bInterval		= 1, /* FIXME: Add this field in the
 				       * HID gadget configuration?
 				       * (struct hidg_func_descriptor)
 				       */
@@ -172,7 +172,7 @@ static struct usb_descriptor_header *hid
 #define CT_FUNC_HID_IDX	0
 
 static struct usb_string ct_func_string_defs[] = {
-	[CT_FUNC_HID_IDX].s	= "HID Interface",
+	[CT_FUNC_HID_IDX].s	= "CDJ-3000 HID",
 	{},			/* end of list */
 };
 
@@ -374,6 +374,11 @@ static void hidg_set_report_complete(str
 	struct f_hidg_req_list *req_list;
 	unsigned long flags;
 
+#ifdef CONFIG_PDJ
+	if (req->status == -ESHUTDOWN)
+		return;
+#endif
+
 	req_list = kzalloc(sizeof(*req_list), GFP_ATOMIC);
 	if (!req_list)
 		return;
@@ -608,11 +613,13 @@ static int hidg_bind(struct usb_configur
 	ep = usb_ep_autoconfig(c->cdev->gadget, &hidg_fs_in_ep_desc);
 	if (!ep)
 		goto fail;
+	ep->desc = &hidg_fs_in_ep_desc;
 	hidg->in_ep = ep;
 
 	ep = usb_ep_autoconfig(c->cdev->gadget, &hidg_fs_out_ep_desc);
 	if (!ep)
 		goto fail;
+	ep->desc = &hidg_fs_out_ep_desc;
 	hidg->out_ep = ep;
 
 	/* preallocate request and buffer */
